working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+
+## Core Library Changes
+
+### Upgrading `dart:async` capabilities
+
+#### Shareable `Future` and `Stream` instances
+
+`Future` and `Stream` should receive the same treatment as `List`: `dart:async`
+should be extended with shareable versions of these and a way to convert an
+existing object to a shareable one.
+
+```dart
+class ShareableFuture<T implements Shareable?>
+    implements Future<T>, Shareable {
+}
+
+class ShareableStream<T implements Shareable?>
+    implements Stream<T>, Shareable {
+}
+
+extension ToShareableFuture<E extends Shareable?> on Future<E> {
+  /// Converts the [Future] to [ShareableFuture] if it is not already shareable.
+  ShareableFuture<E> toShareable() => /* ... */;
+}
+
+extension ToShareableStream<E extends Shareable?> on Stream<E> {
+  /// Converts the [Future] to [ShareableFuture] if it is not already shareable.
+  ShareableStream<E> toShareable() => /* ... */;
+}
+```
+
+The reason for providing these is to allow developers to structure their code
+using well understood primitives futures and streams instead of devising new
+primitives which reimplement some of the same functionality and is compatible
+with threading.
+
+#### Executors of async callbacks
+
+Consider the following code:
+
+```dart
+Future<void> foo() async {
+  await something();
+  print(1);
+}
+
+void main() async {
+  await Isolate.runShared(() async {
+    await foo();
+  });
+}
+```
+
+What happens when `Future` completes in a shared isolate? Who drives event loop
+of that isolate? Which thread will callbacks run on?
+
+I propose to introduce another concept similar to `Zone`: `Executor`. Executors
+encapsulate the notion of the event loop and control how tasks are executed.
+
+```dart
+abstract interface class Executor implements Shareable {
+    /// Returns the current executor.
+    static Executor get current;
+
+    Isolate get owner;
+
+    /// Schedules the given task to run in the given executor.
+    ///
+    /// If the current isolate is not the [owner] of this executor
+    /// the behavior depends on [copy]:   ///
+    ///   * When [copy] is `false` (which is default) [task] is
+    ///     expected to be [Shareable] and is transfered to the
+    ///     target isolate directly. If it is not [Shareable] then
+    ///     `ArgumentError` will be thrown.
+    ///   * If [copy] is `true` then [task] is copied to the
+    ///     target isolate using the same algorithm [SendPort.send]
+    ///     uses.
+    ///
+    void schedule(void Function() task,
+                  {bool copy = false});
+}
+```
+
+How a particular executor runs scheduled tasks depends on the executor itself:
+e.g. an executor can have a pool of threads or notify an embedder that it has
+tasks to run and let embedder run these tasks.
+
+All built-in asynchronous primitives will make the following API guarantee: **a
+callback passed to `Future` or `Stream` APIs will be invoked using executor
+which was running the code which registered the callback.**
```

Comment:
Only shared futures and streams will need that, since a non-shared future can't have listeners in other isolates.

Reply:
Though this is not _exactly_ true. You can actually create unsharable `Future<X>` within shared isolate, which is the only place where you possibly want to have multiple executors. 

As it is written now `ShareableBox` does not actually work in the shared isolate: the provided shim implementation throws `IsolationError` and the actual implementation should probably also throw an error because shared isolate should only have non-concurrent to shareable state. This means the definition of how `fut.then` works also does not work for shared isolate. 

So the question is what happens here:  

```dart
final executor = ThreadPool(concurrency: 2);

executor.schedule(() {
  final Future<List<int>> list = Future.value(<int>[]);
  list..then(cb1)..then(cb2);
});
``` 

This executor can run 2 things concurrently so `cb1` and `cb2` can end up running in parallel. Which is not good. 

I need to think more about how to define things. Maybe it is indeed better to disallow `Future` and `Stream` use outside of proper isolates? Then we also don't need to bother ourselves with event loop and microtasks and all that jazz. In shared isolates if you want to write asynchronous code use coroutines and other abstractions. 

Need to think more on this.

Reply:
An option could be that a shared isolate can recognize that it is one, and not use normal zones. It will automatically run in a root zone which is special, instead of running in zones, it runs in an executor, and any microtask or event creates a new shared isolate to continue the computation. Maybe.
Code inside a shared isolate may still want to create a new zone, which means mutable state